### ISSUE TITLE
Add GNOME Flashback support

### ISSUE THREAD

**Issue by jtojnar** [NixOS/nixpkgs#47451]
We have already packaged GNOME Flashback in #44689, but, we still need a NixOS module to be actually able to run it.

Thanks to [upstream sessions](https://github.com/NixOS/nixpkgs/pull/43992), it is much easier now: see https://github.com/jtojnar/nixpkgs/commits/flashback. Though, we might want to allow enabling Flashback independently from `gnome3` module. Additionally, the upstream session files, unlike the original [packaging effort](https://github.com/NixOS/nixpkgs/pull/26528), do not allow to be easily combined with other window managers like XMonad.

There are also some fixes needed from https://github.com/NixOS/nixpkgs/pull/44497

**Comment by chpatrick** [comment]
Hi, I've managed to get Flashback to work with XMonad. It should also work with the default metacity/compiz but I haven't tried it yet. https://gist.github.com/chpatrick/f7e8c30026634d6344924c87c87a08d6

What do you think is the best way to integrate this?

**Comment by jtojnar** [comment]
> For some strange reason, `gnome-flashback` needs to be added twice like this  `${gnome3.gnome-flashback}/share:${gnome3.gnome-flashback}/share/gsettings-schemas/gnome-flashback-3.28.0` otherwise its gsettings schemas will not be found (according to strace). Do you know why this might be?

Yes, this is normal. The schemas are looked up in `$XDG_DATA_DIRS/glib-2.0/schemas` and we move that directory to `share/gsettings-schemas/$name` to avoid conflicts of `gschemas.compiled`. Normally, `wrapGAppsHook` will take care of this.

**Comment by jtojnar** [comment]
As for the expression:

> `buildDepends`

Those are not necessary, they are not ran during build.

> `XDG_DATA_DIRS`

Yeah, `wrapGAppsHook` to https://github.com/NixOS/nixpkgs/blob/e012ec5b1480f7a76a1509e06ced9c53010b30d9/pkgs/desktops/gnome-3/misc/gnome-flashback/default.nix#L57 will probably do it. Maybe it will even allow to get rid of the rest of the `XDG_DATA_DIRS` setting in `libexec/gnome-flashback-xmonad`.

> `--disable-acceleration-check`

Maybe add a comment why this is necessary.

> `--debug`

should not be necessary, `services.xserver.desktopManager.gnome3.debug` option should enable it.

Otherwise it looks good to me and should be fine for a NixOS module.

**Comment by chpatrick** [comment]
OK, I added `wrapGAppsHook` to the gist. Do you have any thoughts as to how to generalize this to different window managers? Also, should the derivation take the session files from `gnome-flashback` and splice in the given window manager, or should we copy paste them into the derivation?

**Comment by jtojnar** [comment]
> OK, I added wrapGAppsHook to the gist

This will work too, though I was thinking about adding it directly to `gnome-flashback` package.

> Do you have any thoughts as to how to generalize this to different window managers?

You can always parametrize it like https://gist.github.com/jtojnar/1414b2fe9bbfab6253719e985bf2d3a1 but as you noticed, the desktop files are not very elegant. Maybe it will get better when gnome-session switches to systemd services.

> Also, should the derivation take the session files from gnome-flashback and splice in the given window manager, or should we copy paste them into the derivation?

Both have some merit, do not really have a preference.

**Comment by jtojnar** [comment]
We need to take into account that panel applets are not picked up from `XDG_DATA_DIRS` but from `GNOME_PANEL_APPLETS_DIR` environment variable ([source](https://gitlab.gnome.org/GNOME/gnome-panel/blob/b25f16c6077f383a2876512d280f20be10d5ac9a/gnome-panel/libpanel-applet-private/panel-applets-manager-dbus.c#L192-195)). We need to do something similar to what [MATE is doing](https://github.com/NixOS/nixpkgs/blob/0b96dc87406327b715dfc64bfb89f137a4425bbd/nixos/modules/services/x11/desktop-managers/mate.nix#L83).

**Comment by muktupavels** [comment]
@jtojnar That is not correct! `GNOME_PANEL_APPLETS_DIR` only allows to overwrite location and I would say you should not depend on that environment variable...

The real directory is `PANEL_APPLETS_DIR`:
https://gitlab.gnome.org/GNOME/gnome-panel/blob/b25f16c6077f383a2876512d280f20be10d5ac9a/gnome-panel/libpanel-applet-private/Makefile.am#L10
https://gitlab.gnome.org/GNOME/gnome-panel/blob/b25f16c6077f383a2876512d280f20be10d5ac9a/configure.ac#L242

https://gitlab.gnome.org/GNOME/gnome-panel/blob/b25f16c6077f383a2876512d280f20be10d5ac9a/configure.ac#L243
https://gitlab.gnome.org/GNOME/gnome-panel/blob/b25f16c6077f383a2876512d280f20be10d5ac9a/libpanel-applet/libpanel-applet.pc.in#L8

And there are new library that at some point will replace old library!:
https://gitlab.gnome.org/GNOME/gnome-panel/blob/b25f16c6077f383a2876512d280f20be10d5ac9a/gnome-panel/libpanel-applet-private/Makefile.am#L8
https://gitlab.gnome.org/GNOME/gnome-panel/blob/b25f16c6077f383a2876512d280f20be10d5ac9a/gnome-panel/libpanel-applet-private/gp-applet-manager.c#L114

**Comment by jtojnar** [comment]
> The real directory is `PANEL_APPLETS_DIR`:

Yes, I am aware of that. Unfortunately, we cannot use that, since each package on Nix is installed into its own immutable prefix, so we would not be able to add `gnome-applets` or other packages once `gnome-panel` was build.

The only way around it is to patch the program to rely on `XDG_DATA_DIRS` or other environment variable.

**Comment by jtojnar** [comment]
I believe the missing parts are installation of panel applets and removing gnome-shell from gnome-session closure.

**Comment by jtojnar** [comment]
Done

* https://github.com/NixOS/nixpkgs/issues/59344
* https://github.com/NixOS/nixpkgs/issues/67546
